### PR TITLE
Updated coreset subsampling method to improve accuracy

### LINKED ISSUE #67: 1 epoch
Is it meant to be only one epoch in training?
Your config files state 1 epoch, is that just as quick example?
I tried to train PADIM for 10 on mvtec leather and wood and the metrics stay the same anyway, so it seems nothing as gained by training more. 
Lightning module also warn that there is no optimizer so I guess train only finds the correct thresholds and that takes 1 epoch.

### LINKED ISSUE #74: Diverged metrics in PatchCore: dropping from 0.99 to 0.44 and 0.03 is rather critical?
diverged metrics: dropping from 0.99 to 0.44 and 0.03 is rather critical?
log images: nice :) 
Also padim dropped in performance, but not as crazy.
Here's a patchcore result of "good" parts: 

this is padim:
DATALOADER:0 TEST RESULTS
{'image_AUROC': 0.7589669823646545,
 'image_F1': 0.8787878751754761,
 'pixel_AUROC': 0.9781586527824402,
 'pixel_F1': 0.22379672527313232}
008padim

_Originally posted by @sequoiagrove in https://github.com/openvinotoolkit/anomalib/issues/67#issuecomment-1010091154_

### PR DESCRIPTION
**Pull request by blakshma**
Fixes #74 and #6

### CODE REVIEW

File: anomalib/models/patchcore/config.yaml [hunk 1]
```diff
@@ -14,7 +14,7 @@ dataset:
     use_random_tiling: False
     random_tile_count: 16
   image_size: 224
-  train_batch_size: 32
+  train_batch_size: 1
```
Comment (ashwinvaidya17) [inline anomalib/models/patchcore/config.yaml:17]: Is this by choice or an artefact of debugging?

File: anomalib/models/patchcore/model.py [hunk 2]
```diff
@@ -311,7 +289,55 @@ def validation_step(self, batch, _):  # pylint: disable=arguments-differ
...
+        eps: float = 0.90,
+        force_cpu: bool = False,
+    ) -> Tensor:
+        """Returns n subsampled coreset idx for given sample_set.
+
+        Args:
+            embedding:      (sample_count, d) tensor of patches.
+            sample_count:          Number of patches to select.
+            eps:        Parameter for spare projection aggresion.
+            force_cpu:  Force cpu, useful in case of GPU OOM.
```
Comment (ashwinvaidya17) [inline anomalib/models/patchcore/model.py:311]: These seem to be missing the types.

**Review by ashwinvaidya17 — APPROVED** [review]
Thanks for addressing this! It's also nice that we could get rid of all the extra modules. I have only two minor comments.

**Review by samet-akcay — COMMENTED** [review]
Forgot to submit my review

File: anomalib/models/patchcore/config.yaml (hunk 1 above)
Comment (blakshma) [reply anomalib/models/patchcore/config.yaml]: Good catch. It was an artifact. Fixed it.

[the remark this answers is not in the document]
Comment (blakshma) [reply anomalib/models/patchcore/config.yaml]: Good catch. It was an artifact. Fixed it.

File: anomalib/models/patchcore/model.py [hunk 3]
```diff
@@ -311,7 +289,55 @@ def validation_step(self, batch, _):  # pylint: disable=arguments-differ
...
+            min_distances = min_distances.to("cuda")
+
+        for _ in range(sample_count - 1):
+            distances = torch.linalg.norm(sample_set - last_item, dim=1, keepdims=True)  # broadcast
+            min_distances = torch.minimum(distances, min_distances)  # iterate
+            select_idx = torch.argmax(min_distances)  # select
+
+            last_item = sample_set[select_idx : select_idx + 1]
+            min_distances[select_idx] = 0
+            coreset_idx.append(select_idx.to("cpu"))  # type: ignore[attr-defined]
```
[the remark this answers is not in the document]
Comment (blakshma) [reply anomalib/models/patchcore/model.py]: It was due to the way select_idx was initialized. Updated this in the latest commit.

**Comment by blakshma** [comment]
Thanks for the review @samet-akcay @ashwinvaidya17 . I have addressed them in the recent commit.

**Review by djdameln — COMMENTED** [review]
@blakshma Thanks a lot for this PR. Great to finally see the Patchcore implementation produce results that are in line with the paper. Also nice to get rid of the helper classes, the implementation is much cleaner this way. Just two minor comments.

File: anomalib/models/patchcore/model.py [hunk 4]
```diff
@@ -311,7 +289,53 @@ def validation_step(self, batch, _):  # pylint: disable=arguments-differ
...
         """
 
-        anomaly_maps, _ = self.model(batch["image"])
+        anomaly_maps, anomaly_score = self.model(batch["image"])
         batch["anomaly_maps"] = anomaly_maps
+        batch["pred_scores"] = anomaly_score.unsqueeze(0)
 
         return batch
+
+    def get_coreset_idx(
```
[the remark this answers is not in the document]
Comment (blakshma) [reply anomalib/models/patchcore/model.py]: That's a great idea. I have made that change.

**Review by djdameln — APPROVED** [review]
